### PR TITLE
feat: flyway module

### PR DESCRIPTION
**Pull request by mdelapenya**
I tried to include one single change for each commit, so I recommend you visiting each one in order to understand the mental process to build the module.

PLMK if you need more context :pray:

- **chore: add golangci-lint support**
- **chore: update makefile to run tests**
- **chore: run make lint**
- **chore: ignore test output files**
- **fix: use proper network setup**
- **chore: simplify calculation of the internal URL**
- **chore: rename to a more meaningful name**
- **chore: use filepath for proper OS paths**
- **chore: rename to testdata**
- **fix: let users pass migrations**
- **chore: rename variables removing package namespace**
- **fix: proper value for exposed const**
- **chore: use testcontainers WithEnv**
- **chore: simplify option names**
- **chore: rename to migrations**
- **chore: include all options in example**
- **feat: more consistent wait for strategy**
- **chore: support passing a timeout**
- **chore: parse request for validation**

### CODE REVIEW

File: flyway.go [hunk 1]
```diff
@@ -36,25 +48,20 @@ type FlywayContainer struct {
...
-			{
-				HostFilePath:      fmt.Sprintf("./test%s", defaultFlywayMigrationsPath),
-				ContainerFilePath: defaultFlywayMigrationsPath,
-			},
+			flywayEnvLocationsKey:      fmt.Sprintf("filesystem:%s", DefaultMigrationsPath),
 		},
 		Cmd: []string{
 			migrateCmd, infoCmd,
 		},
+		WaitingFor: wait.ForAll(
```
Comment (KevinJTaylor) [inline flyway.go:60]: I was wondering why you'd not used the WithTimeout func below, but I see now its different ... I sometimes provide a builder in such cases (not tested this), it just saves having a similar snippet in both places

```
func buildWait(timeout Duration) {
  wait.ForAll(
     wait.ForExit().WithExitTimeout(timeout),
     waitForApplied,
     waitForValidated,
  )
}
```

File: flyway.go [hunk 2]
```diff
@@ -68,6 +75,10 @@ func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomize
 		}
 	}
 
+	if err := parseRequest(genericContainerReq); err != nil {
+		return nil, err
+	}
```
Comment (KevinJTaylor) [inline flyway.go:80]: this is just a language thing I guess, but a parser to me is something which takes an input and parses it (converts) it to something, whereas here what we're really doing is validating the request. If it was a boolean response I tend to prefix the func with is/was, but here as it is an error result, I would generally do

```
	if err := validateRequest(genericContainerReq); err != nil {
		return nil, err
	}
```

NOTE: I don't mind really and if you prefer parse we can leave it as is

File: flyway.go [hunk 3]
```diff
@@ -88,66 +99,99 @@ func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomize
...
+
+	return nil
+}
+
+func WithUser(user string) testcontainers.CustomizeRequestOption {
 	return withEnvSetting("FLYWAY_USER", user)
 }
 
-func WithEnvPassword(password string) testcontainers.CustomizeRequestOption {
+func WithPassword(password string) testcontainers.CustomizeRequestOption {
```
Comment (KevinJTaylor) [inline flyway.go:143]: With these type of funcs I tend to think that the name reflects what you're changing ... by including Env in the name I wanted it to be clear it was an environment setting, again not significant really and happy to go with your choice

File: flyway.go [hunk 4]
```diff
@@ -88,66 +99,99 @@ func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomize
...
-func WithEnvLocations(locations string) testcontainers.CustomizeRequestOption {
-	return withEnvSetting("FLYWAY_LOCATIONS", locations)
-}
-
 func withEnvSetting(key, group string) testcontainers.CustomizeRequestOption {
-	return func(req *testcontainers.GenericContainerRequest) error {
-		req.Env[key] = group
-		return nil
-	}
+	return testcontainers.WithEnv(map[string]string{
```
Comment (KevinJTaylor) [inline flyway.go:176]: ahhh, this is what you meant, good

File: flyway.go [hunk 5]
```diff
@@ -88,66 +99,99 @@ func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomize
...
+		key: group,
+	})
 }
 
-func WithNetwork(network string) testcontainers.CustomizeRequestOption {
+func WithMigrations(absHostFilePath string) testcontainers.CustomizeRequestOption {
```
Comment (KevinJTaylor) [inline flyway.go:181]: I think this is the one change I was unsure about, but again its not a very significant thing. I was allowing the caller of the service to define the path inside the container to the migrations, but by doing this you are fixing them to the default path.

NOTE: I'm pretty sure I hit problems previously where the local path had to be follow the same convention, so may be I need to test that too e.g. .../flyway/sql

File: flyway_test.go [hunk 6]
```diff
@@ -30,12 +34,8 @@ type intPostgresContainer struct {
...
 }
 
-func (c *intPostgresContainer) getInternalUrl(ctx context.Context) (string, error) {
-	json, err := c.Inspect(ctx)
-	if err != nil {
-		return "", err
-	}
-	return fmt.Sprintf("jdbc:postgresql:/%s:%s/%s", json.Name, defaultPostgresPort, defaultPostgresDbName), nil
+func (c *intPostgresContainer) getNetworkUrl() string {
+	return fmt.Sprintf("jdbc:postgresql://%s:%s/%s?sslmode=disable", defaultPostgresSrvName, defaultPostgresPort, defaultPostgresDbName)
```
Comment (KevinJTaylor) [inline flyway_test.go:38]: like

File: flyway_test.go [hunk 7]
```diff
@@ -83,15 +83,70 @@ func TestFlyway(t *testing.T) {
 	require.Equal(t, 0, state.ExitCode, "container exit code was not as expected: migration failed")
 }
 
-func createTestNetwork(ctx context.Context) (*testcontainers.DockerNetwork, error) {
-	return tcnetwork.New(ctx)
+func TestFlyway_parseInvalidReques(t *testing.T) {
```
Comment (KevinJTaylor) [inline flyway_test.go:86]: good

**Review by KevinJTaylor — APPROVED** [review]

File: flyway.go (hunk 1 above)
Comment (mdelapenya) [reply flyway.go:60]: Indeed. I did not want to make it more complicated using an Options struct that satisfies the RequestCustomiser struct (please see r[edpanda's options](https://github.com/testcontainers/testcontainers-go/blob/main/modules/redpanda/options.go#L48-L71) and [how they are evaluated](https://github.com/testcontainers/testcontainers-go/blob/main/modules/redpanda/redpanda.go#L89-L95)): we could have defined the timeout as a field in that Options struct, and before running the container building the waitFor just once for the given options.

File: flyway.go (hunk 2 above)
Comment (mdelapenya) [reply flyway.go:80]: I've used validated in the past, but starting to see parse more and more as a way to check for errors. Either name is fine for me, so feel free to change it to the one you prefer.

File: flyway.go (hunk 3 above)
Comment (mdelapenya) [reply flyway.go:143]: But the module should hide the implementation details, so users should not care about if it's an env, a config file, or any other thing. We should probably add comments for each option.

File: flyway.go (hunk 5 above)
Comment (mdelapenya) [reply flyway.go:181]: I'd advocate for having opinionated modules that offers a deep API to configure them. If flyway expects the migrations in one place, why adding complexity and allowing the user to change it? In the end what it's important is the user files, not the container files.

**Pull request merged**